### PR TITLE
#1603 Add configuration to override localstack image location

### PR DESCRIPTION
**Pull request by wavetylor**
Adds support for property`localstack.container.image` in `testcontainer.properties`. 

Also moves the default `VERSION` in LocalStackContainer to be read from TestcontainersConfiguration.

### CODE REVIEW

File: core/src/main/java/org/testcontainers/utility/TestcontainersConfiguration.java [hunk 1]
```diff
@@ -145,4 +153,18 @@ private static TestcontainersConfiguration loadConfiguration() {
...
         return config;
     }
+
+    /**
+     * <p>Only for testing to workaround the limitations of static, lazy singletons enforced by lombok</p>
+     *
+     * @param directoryForGlobalProps if null, defaults to {@code System.property("user.home") }
+     */
+    @VisibleForTesting
+    public void resetConfiguration(@Nullable String directoryForGlobalProps) {
```
Comment (bsideup) [inline core/src/main/java/org/testcontainers/utility/TestcontainersConfiguration.java]: please use system properties for testing, not the files

File: modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java [hunk 2]
```diff
@@ -28,16 +29,15 @@
  */
 public class LocalStackContainer extends GenericContainer<LocalStackContainer> {
 
-    public static final String VERSION = "0.8.6";
 
     private final List<Service> services = new ArrayList<>();
 
     public LocalStackContainer() {
-        this(VERSION);
+        this(TestcontainersConfiguration.getInstance().getLocalStackImage().split(":")[1]);
```
Comment (bsideup) [inline modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java]: by default, we should use the provided image

File: modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java [hunk 3]
```diff
@@ -28,16 +29,15 @@
...
     }
 
     public LocalStackContainer(String version) {
-        super("localstack/localstack:" + version);
+        super(TestcontainersConfiguration.getInstance().getLocalStackImage().split(":")[0] + ":" + version);
```
Comment (bsideup) [inline modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java]: if the user specifies the version, we should use the official image

File: modules/localstack/src/test/java/org/testcontainers/containers/localstack/LocalStackImageTest.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import java.io.IOException;
+
+public class LocalStackImageTest {
+
+    @Rule
+    public TemporaryFolder folder = new TemporaryFolder();
+
+    @Before
+    public void setupConfig() throws IOException {
+        File tempDir = folder.newFolder();
```
Comment (bsideup) [inline modules/localstack/src/test/java/org/testcontainers/containers/localstack/LocalStackImageTest.java]: please use system properties instead

File: modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java (hunk 3 above)
Comment (wavetylor) [reply modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java]: This would use the default version, unless the user has already overridden the property.

File: modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java (hunk 2 above)
Comment (wavetylor) [reply modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java]: This would use the default version, unless the user has already overridden the property, which is then exactly what this PR is addressing.

File: core/src/main/java/org/testcontainers/utility/TestcontainersConfiguration.java (hunk 1 above)
Comment (wavetylor) [reply core/src/main/java/org/testcontainers/utility/TestcontainersConfiguration.java]: Can you help me understand what you mean here? As far as I can tell, the only way to change these properties is to use files.

File: modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java (hunk 3 above)
Comment (bsideup) [reply modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java]: > This would use the default version, unless the user has already overridden the property.

Yes.

However, we should use the official version, otherwise it breaks the expectation if the overridden image does not have that image tagged

File: modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java (hunk 2 above)
Comment (bsideup) [reply modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java]: If the user overrides the image, we should use it "as it" and pass to `super(TestcontainersConfiguration.getInstance().getLocalStackImage())`.

The image might not have `:` at all

File: core/src/main/java/org/testcontainers/utility/TestcontainersConfiguration.java (hunk 1 above)
Comment (bsideup) [reply core/src/main/java/org/testcontainers/utility/TestcontainersConfiguration.java]: Sorry, my bad, I was thinking about another project :D

We need to think about the testing strategy here... a public method is definitely a no go

Comment (wavetylor) [reply core/src/main/java/org/testcontainers/utility/TestcontainersConfiguration.java]: like my comment says, this class has serious testing limitations. `@VisibleForTesting` is a pretty common pattern here.

File: modules/localstack/src/test/java/org/testcontainers/containers/localstack/LocalStackImageTest.java (hunk 4 above)
Comment (wavetylor) [reply modules/localstack/src/test/java/org/testcontainers/containers/localstack/LocalStackImageTest.java]: I assume you were thinking of the other project here as well?

File: core/src/main/java/org/testcontainers/utility/TestcontainersConfiguration.java [hunk 5]
```diff
@@ -18,8 +21,9 @@
 public class TestcontainersConfiguration {
 
     private static String PROPERTIES_FILE_NAME = "testcontainers.properties";
+    private static String SYSTEM_PROPERTY_USER_HOME = "user.home";
 
-    private static File GLOBAL_CONFIG_FILE = new File(System.getProperty("user.home"), "." + PROPERTIES_FILE_NAME);
+    private static File GLOBAL_CONFIG_FILE = new File(System.getProperty(SYSTEM_PROPERTY_USER_HOME), "." + PROPERTIES_FILE_NAME);
```
Comment (bsideup) [inline core/src/main/java/org/testcontainers/utility/TestcontainersConfiguration.java]: unrelated change, could you please revert it?

File: core/src/main/java/org/testcontainers/utility/TestcontainersConfiguration.java (hunk 1 above)
Comment (bsideup) [reply core/src/main/java/org/testcontainers/utility/TestcontainersConfiguration.java]: we should definitely improve the testing support for this class, but we cannot afford having this method as a public method

File: docs/features/configuration.md [hunk 6]
```diff
@@ -42,6 +42,9 @@ Some companies disallow the usage of Docker Hub, but you can override `*.image`
 > **kafka.container.image = confluentinc/cp-kafka**  
 > Used by KafkaContainer 
 
+> **localstack.container.image = localstack/localstack:0.8.6**  
+> Use by LocalStackContainer
```
Comment (bsideup) [inline docs/features/configuration.md]:
```suggestion
> Used by LocalStackContainer
```

**Comment by rnorth** [comment]
Closing, as #1873 proceeded to make this configurable.
